spirv_hlsl.cpp

```diff
@@ -3598,6 +3598,18 @@ string CompilerHLSL::bitcast_glsl_op(const SPIRType &out_type, const SPIRType &i
 		}
 		return "spvPackFloat2x16";
 	}
+	else if (out_type.basetype == SPIRType::UShort && in_type.basetype == SPIRType::Half)
+	{
+		if (hlsl_options.shader_model < 50)
+			SPIRV_CROSS_THROW("Half to UShort requires Shader Model 5.");
```

Comment:
This requires SM 6.2 I think.

Reply:
Both `f16tof32()` and `f32tof16()` are actually supported from as low as SM 4.0 (via emuation; natively from 5.0). Therefore, this code (generated by SPIRV-Cross using SM 4.0) compiles on DXC with just `-T cs_4_0` (the rest of the file is omitted for convenience):
```
min16uint AW1_x(min16uint a)
{
    return a;
}

min16float APrxLoRsqH1(min16float a)
{
    min16uint param = 22947u;
    min16uint param_1 = 1u;
    return f16tof32(AW1_x(param) - (f32tof16(a) >> AW1_x(param_1)));
}
```

So, I've changed the check to be happy even with SM 4.0 for now.

Reply:
It's not f16tof32 I'm concerned about, it's UShort and Half types. These are explicit 16-bit types in SPIR-V and cannot be supported properly before SM 6.2. min16float/min16int is basically RelaxedPrecision.

Reply:
To be honest, I'm not sure I undersand the full implications of that. I can just change the check to the following and so be safe:
```
	if (!hlsl_options.enable_16bit_types) // Implies SM 6.2
		SPIRV_CROSS_THROW("[...] to [...] requires native 16-bit type support.");
```

What do you think?

Reply:
Yes, that looks better. I forgot there was a enable_16bit_types option in SPIRV-Cross, but yes. Bitcasts in HLSL on min16float types would fail in DXC, so this is more correct.

Reply:
Just in case, before making the change, in the current state of this PR SPIRV-Cross can generate this (with SM 4.0 and 16-bit types disabled):
```
min16uint AW1_x(min16uint a)
{
    return a;
}

min16float APrxLoRsqH1(min16float a)
{
    min16uint param = 22947u;
    min16uint param_1 = 1u;
    return (min16float)f16tof32(AW1_x(param) - ((min16uint)f32tof16(a) >> AW1_x(param_1)));
}

min16float APrxLoRcpH1(min16float a)
{
    min16uint param = 30596u;
    return (min16float)f16tof32(AW1_x(param) - (min16uint)f32tof16(a));
}
```

And it compiles fine on DXC. No warnings, no errors. And the shader works fine at runtime.